lldb/source/Plugins/SymbolLocator/Debuginfod/SymbolLocatorDebuginfod.cpp

```diff
@@ -114,29 +177,85 @@ SymbolLocator *SymbolLocatorDebuginfod::CreateInstance() {
 
 static std::optional<FileSpec> GetFileForModule(
     const ModuleSpec &module_spec,
-    std::function<llvm::Expected<std::string>(llvm::object::BuildIDRef)>
-        PullFromServer) {
+    std::function<llvm::StringRef(llvm::object::BuildID)> UrlBuilder,
+    bool force_lookup) {
+  PluginProperties &plugin_props = GetGlobalPluginProperties();
+  llvm::Expected<llvm::StringRef> CacheDirectoryPathOrErr =
+      plugin_props.GetCachePath();
+  // We really do require a cache location to store stuff.
+  if (!CacheDirectoryPathOrErr)
+    return {};
+  llvm::StringRef CacheDirectoryPath = *CacheDirectoryPathOrErr;
+  llvm::SmallVector<llvm::StringRef> DebuginfodUrls =
+      llvm::getDefaultDebuginfodUrls();
+  std::chrono::milliseconds Timeout = plugin_props.GetTimeout();
   if (!ModuleList::GetGlobalModuleListProperties().GetEnableExternalLookup())
     return {};
   const UUID &module_uuid = module_spec.GetUUID();
   if (module_uuid.IsValid() && llvm::canUseDebuginfod()) {
     llvm::object::BuildID build_id(module_uuid.GetBytes());
-    llvm::Expected<std::string> result = PullFromServer(build_id);
+    llvm::StringRef UrlPath = UrlBuilder(build_id);
+    std::string CacheKey = llvm::getDebuginfodCacheKey(UrlPath);
+    bool ask_server = force_lookup || plugin_props.GetLookupMode() ==
+                                          SymbolLookupMode::eLookupModeAlways;
```

Comment:
Do we not have a setting to specify async download? Seems like it would be nice to be able to download in the background for system libraries that we do end up downloading. This forces us to either run a user command or specify we always want symbols for everything, and if we do that, then this becomes synchronous.

Reply:
I'm working on tests to validate this, but if my memory from last month serves, the "force_lookup" (now sync_lookup) makes on-demand do everything asychronously unless the user specifically performed an action that triggered symbol download. If that's not the case, I'll add an "async" option to the SymbolLookupMode enum.